### ISSUE TITLE
Shuffle mode is unable to reverse to the previous song

### ISSUE THREAD

**Issue by scresante** [quodlibet/quodlibet#3592]
Steps to reproduce
------------------

Enter shuffle mode. 
Skip to the next song. 
skip previous: "Not playing" comes up, no song plays.

Expected Output
---------------

The last song shuffled should play
Quodlibet should keep a list of the last tracks played for all the modes. and store these lists and consistently revert to the last played song in that mode.

Actual Output
-------------

No logs yet.

Test System
-----------

### Which version of Quod Libet?

4.4.0 

### Which operating system 

Archlinux

I thought this was fixed already, this is a pretty basic UX problem.

**Comment by mwlsyn** [comment]
More information:

This is a bug which appears in most of the views in the Browse menu, but for Playlists the bug is not reproduced on my system (skip previous works in that case).

This bug completely breaks shuffle behavior because the history of played songs is eliminated, so there is nothing to stop previously-played songs in the list from being selected as the next song. This is especially bad for album views or filter views with small numbers of songs.

**Comment by declension** [comment]
Can anyone track down when exactly it stopped working? That would help a lot.

**Comment by bassstorm** [comment]
@declension I could try during weekend, will keep you posted.

**Comment by bassstorm** [comment]
@declension et al,
I was able to reproduce the issue on `release-4.4.0`. But good news it works properly on latest `master`. Specifically it was fixed by [that commit](https://github.com/quodlibet/quodlibet/commit/68ba6786072557d05f9da99c25d14d1d8cf1d192).

**Comment by declension** [comment]
Great work thanks @bassstorm !

@scresante are you comfortable with running a [dev build](https://quodlibet.readthedocs.io/en/latest/development/devenv.html) to confirm it's fixed in `master`?

**Comment by scresante** [comment]
For the sake of simplicity I'd say that the elegant and simple that @bassstorm  made definitely solves this bug. Thanks!

I am fine with forward always shuffling. That makes more sense. If I wanted something else, I'd write my own playlist auto-logger. (The idea being random shuffle sometimes comes up with brilliant playlists).

**Comment by dannn-o** [comment]
Encountered this with new Linux Mint 20.1 box running Quod Libet 4.4 (and can confirm that Quod Libet 4.3 on my Mint 19.3 box did **not** have the issue).  

I didn't see instructions at ReadTheDoc for compiling, and can be patient to wait for the next release from the PPA maintained by @lazka (ppa:lazka/ppa) if it will be in the somewhat near future.  Can anyone advise to an expected date for that, or offer a (simpler than compiling) workaround for the issue?  I didn't realize just how much I like to reverse back to the previous song while in shuffle mode.  Thanks in advance.

**Comment by lual** [comment]
@dannn-o  I do not recommend this, but this is Python: No need to compile something. Just implement [this commit ](https://github.com/quodlibet/quodlibet/commit/68ba6786072557d05f9da99c25d14d1d8cf1d192) as monkey patch. Therefore you have to find the sourcecode and edit the file as root-user.
On my system /usr/lib/python3.9/site-packages/quodlibet/qltk/songlist.py
Just be aware: [The Python Code Indentation](https://en.wikipedia.org/wiki/Python_syntax_and_semantics#Indentation). Keep the replacement if-statement aligned with the original!

**Comment by dannn-o** [comment]
@lual I finally had some time this morning to implement the patch, but found that Quod Libet hangs and will not launch with your code added.  The first line has 3 parameters for the GetBoolean function, and I'm not positive but I think it only expects 2:

`if not config.getboolean("memory", "shuffle", False) and \`

Unless you meant "raw=False"?

**Comment by lual** [comment]
@dannn-o 
The mentioned commit is not my code, but i've tried it. `getboolean` parameters are: section, key and a fallback-value (if no value was found in config). It should work.
Maybe you messed up the indentation? Try to start `quodlibet` from command line and have a look on the error messages.

Anyway - i hope a fix for this bug will make it into the upcoming Release 4.4.1

**Comment by dannn-o** [comment]
Careful as I tried to be, I guess I didn't get the indentation exactly right.  I had even tried combining the lines into one, pasting it right over the original source file:
`if config.getboolean("song_list", "auto_sort") and self.is_sorted() and not config.getboolean("memory", "shuffle", False):`

But, command line (as you suggested) kicked out an "unexpected indent" as it somehow got to be one space off.  Once I corrected that, the fix worked (although, I guess I have to go reread the documentation to understand the 3rd parameter usage as it wasn't explained well).  That feature (moving backward in a shuffle playlist) is one I use often, so the patch should server to keep me going until the 4.4.1 release officially hits the repos.
